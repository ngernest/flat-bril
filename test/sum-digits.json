{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v0"
          ],
          "dest": "total",
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.1"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v2",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v3",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v2",
            "v3"
          ],
          "dest": "v4",
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "v4"
          ],
          "labels": [
            "for.body.1",
            "for.end.1"
          ],
          "op": "br"
        },
        {
          "label": "for.body.1"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v5",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v6",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "v5",
            "v6"
          ],
          "dest": "v7",
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v7"
          ],
          "dest": "tenth",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v8",
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v9",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "tenth"
          ],
          "dest": "v10",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v9",
            "v10"
          ],
          "dest": "v11",
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v8",
            "v11"
          ],
          "dest": "v12",
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v12"
          ],
          "dest": "currentDigit",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "total"
          ],
          "dest": "v13",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "currentDigit"
          ],
          "dest": "v14",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v13",
            "v14"
          ],
          "dest": "v15",
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v15"
          ],
          "dest": "total",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "tenth"
          ],
          "dest": "v16",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v16"
          ],
          "dest": "n",
          "op": "id",
          "type": "int"
        },
        {
          "labels": [
            "for.cond.1"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.1"
        },
        {
          "args": [
            "total"
          ],
          "dest": "v17",
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v17"
          ],
          "op": "print"
        },
        {
          "dest": "v18",
          "op": "const",
          "type": "int",
          "value": 0
        }
      ],
      "name": "main"
    }
  ]
}
