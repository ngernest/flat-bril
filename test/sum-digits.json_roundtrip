{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v0"
          ],
          "dest": "total",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.1"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v2",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v3",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "v2",
            "v3"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [
            "for.body.1",
            "for.end.1"
          ],
          "op": "br"
        },
        {
          "label": "for.body.1"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v5",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v6",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "v5",
            "v6"
          ],
          "dest": "v7",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "v7"
          ],
          "dest": "tenth",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "n"
          ],
          "dest": "v8",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v9",
          "op": "const",
          "type": "int",
          "value": 10
        },
        {
          "args": [
            "tenth"
          ],
          "dest": "v10",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v9",
            "v10"
          ],
          "dest": "v11",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "v8",
            "v11"
          ],
          "dest": "v12",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "v12"
          ],
          "dest": "currentDigit",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "total"
          ],
          "dest": "v13",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "currentDigit"
          ],
          "dest": "v14",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v13",
            "v14"
          ],
          "dest": "v15",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "v15"
          ],
          "dest": "total",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "tenth"
          ],
          "dest": "v16",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v16"
          ],
          "dest": "n",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.1"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end.1"
        },
        {
          "args": [
            "total"
          ],
          "dest": "v17",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v17"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "dest": "v18",
          "op": "const",
          "type": "int",
          "value": 0
        }
      ],
      "name": "main"
    }
  ]
}
