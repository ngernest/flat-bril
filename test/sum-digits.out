@main
	op: const		dest: v0		type: int		value: 0	
	op: id   		dest: total		type: int		args: ["v0"]	
.for.cond.1
	op: id   		dest: v2		type: int		args: ["n"]	
	op: const		dest: v3		type: int		value: 0	
	op: gt   		dest: v4		type: bool		args: ["v2", "v3"]	
	op: br   		args: ["v4"]		labels: ["for.body.1", "for.end.1"]
.for.body.1
	op: id   		dest: v5		type: int		args: ["n"]	
	op: const		dest: v6		type: int		value: 10	
	op: div  		dest: v7		type: int		args: ["v5", "v6"]	
	op: id   		dest: tenth		type: int		args: ["v7"]	
	op: id   		dest: v8		type: int		args: ["n"]	
	op: const		dest: v9		type: int		value: 10	
	op: id   		dest: v10		type: int		args: ["tenth"]	
	op: mul  		dest: v11		type: int		args: ["v9", "v10"]	
	op: sub  		dest: v12		type: int		args: ["v8", "v11"]	
	op: id   		dest: currentDigit		type: int		args: ["v12"]	
	op: id   		dest: v13		type: int		args: ["total"]	
	op: id   		dest: v14		type: int		args: ["currentDigit"]	
	op: add  		dest: v15		type: int		args: ["v13", "v14"]	
	op: id   		dest: total		type: int		args: ["v15"]	
	op: id   		dest: v16		type: int		args: ["tenth"]	
	op: id   		dest: n 		type: int		args: ["v16"]	
	op: jmp  		labels: ["for.cond.1"]
.for.end.1
	op: id   		dest: v17		type: int		args: ["total"]	
	op: print		args: ["v17"]	
	op: const		dest: v18		type: int		value: 0	
